[C++ API] Add functional overloads for pixelshuffle, pooling, upsampling, vision

yf225 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29360 Add functional overloads for fold, linear, loss, normalization, padding
* **#29359 Add functional overloads for pixelshuffle, pooling, upsampling, vision**
* #29358 Add functional overloads for activation, batchnorm, distance, embedding

This PR adds functional overloads that take the full set of arguments (instead of just Options) for the following functionals:
- pixelshuffle
- pooling
- upsampling
- vision

These new functionals lives in the `torch::nn::functional::detail` namespace and they are only meant to be called from the module forward methods (i.e. they are not public API). This is in preparation for the future change where we make module Options and functional Options two different classes, because if the module forward method has to construct a new functional Options object every time it runs it will be pretty silly and bad performance.

Differential Revision: [D18376978](https://our.internmc.facebook.com/intern/diff/D18376978)

Commit:
Add functional overloads for pixelshuffle, pooling, upsampling, vision

Commit:
Update on "Add functional overloads for pixelshuffle, pooling, upsampling, vision"

Add functional overloads for pixelshuffle, pooling, upsampling, vision

gh-metadata: pytorch pytorch 29359 gh/yf225/38/head

Commit:
Update on "Add functional overloads for pixelshuffle, pooling, upsampling, vision"

Add functional overloads for pixelshuffle, pooling, upsampling, vision

gh-metadata: pytorch pytorch 29359 gh/yf225/38/head

kostmo commented:
## CircleCI build failures summary
As of commit e12d49a:
* 0/1 recognized as flaky
* 1/1 failures introduced in this PR



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=e12d49a6d75f6a102d8c49302d4fc5a683b2c37c).

---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

pbelevich commented on torch/csrc/api/include/torch/nn/functional/vision.h:
```diff
@@ -47,52 +47,70 @@ inline Tensor affine_grid(
     const Tensor& input,
     const Tensor& grid,
-    GridSampleOptions options = {}) {
+    std::string mode = "bilinear",
```
Are we going to change it to torch:: enums?

yf225 replied:
Yes I will do it in a follow-up PR (before the release)

pbelevich commented on torch/csrc/api/include/torch/nn/functional/vision.h:
```diff
@@ -47,52 +47,70 @@ inline Tensor affine_grid(
+inline Tensor grid_sample(
+    const Tensor& input,
+    const Tensor& grid,
+    GridSampleOptions options = {}) {
```
why by value?

yf225 replied:
replied in an earlier comment

pbelevich commented:
If `torch::nn::functional::detail` methods are supposed to be called by forward methods and are not part of public API why do they have default values?

Commit:
Update on "[C++ API] Add functional overloads for pixelshuffle, pooling, upsampling, vision"

Add functional overloads for pixelshuffle, pooling, upsampling, vision

gh-metadata: pytorch pytorch 29359 gh/yf225/38/head

Commit:
Update on "[C++ API] Add functional overloads for pixelshuffle, pooling, upsampling, vision"

Add functional overloads for pixelshuffle, pooling, upsampling, vision

gh-metadata: pytorch pytorch 29359 gh/yf225/38/head

Commit:
Update on "[C++ API] Add functional overloads for pixelshuffle, pooling, upsampling, vision"

Add functional overloads for pixelshuffle, pooling, upsampling, vision

gh-metadata: pytorch pytorch 29359 gh/yf225/38/head

Commit:
Update on "[C++ API] Add functional overloads for pixelshuffle, pooling, upsampling, vision"

Add functional overloads for pixelshuffle, pooling, upsampling, vision

gh-metadata: pytorch pytorch 29359 gh/yf225/38/head

yf225 commented on torch/csrc/api/include/torch/nn/functional/upsampling.h:
```diff
@@ -42,65 +48,75 @@ inline Tensor interpolate(const Tensor& input, InterpolateOptions options) {
 }
+} // namespace detail
+
+inline Tensor interpolate(const Tensor& input, InterpolateOptions options = {}) {
```
In some of the functionals we want to be able to modify the fields in the options (such as `out` in `NormalizeFuncOptions` added in https://github.com/pytorch/pytorch/pull/29404), and Ed also suggested that we should just pass options by value because it's a small struct (https://github.com/pytorch/pytorch/pull/27942#discussion_r337604093).

yf225 commented on torch/csrc/api/include/torch/nn/functional/upsampling.h:
```diff
@@ -9,30 +9,36 @@ namespace torch {
+namespace detail {
+inline Tensor interpolate(
+  const Tensor& input,
+  std::vector<int64_t> size = {},
```
Fixed

yf225 commented:
> If `torch::nn::functional::detail` methods are supposed to be called by forward methods and are not part of public API why do they have default values?

Thanks a lot for the suggestion - I removed the default values from all of them.
